federationclient.go

```diff
@@ -110,7 +110,27 @@ func (ac *FederationClient) SendJoin(
 	if err = req.SetContent(event); err != nil {
 		return
 	}
-	err = ac.doRequest(ctx, req, &res)
+	var intermediate struct {
+		StateEvents []json.RawMessage `json:"state"`
+		AuthEvents  []json.RawMessage `json:"auth_chain"`
+		Origin      ServerName        `json:"origin"`
+	}
+	process := func() {
+		for _, se := range intermediate.StateEvents {
+			if ev, err := NewEventFromUntrustedJSON(se, roomVersion); err == nil {
+				res.StateEvents = append(res.StateEvents, ev)
```

Comment:
Given how large these arrays can be, maybe allocate a capacity to `len(intermediate.StateEvents)` to avoid too much churn?

Reply:
Good point, done.